File: Yr2022/day_three.py
from time import perf_counter_ns
from typing import List

def get_common_item(input_rucksack: str):
    first, second = input_rucksack[:len(input_rucksack)//2], input_rucksack[len(input_rucksack)//2:]
    return ''.join( set(first).intersection(set(second)) )

def get_common_item_group(input_list : List):
    first, second, third = input_list[0], input_list[1], input_list[2]
    return ''.join( set(first).intersection(set(second)).intersection(set(third)) )

def get_priority(input_char: str):
    ord_shift = 38 if input_char.isupper() else 96
    return ord(input_char) - ord_shift

def calculate_total_priority(filename):
    priority_counter = 0
    badge_priority_counter = 0 
    group_list = []
    with open(filename, 'r') as file:
        for line in file:
            priority_counter += get_priority(get_common_item(line))
            group_list.append(line.strip())
            if len(group_list) == 3:
                badge_priority_counter += get_priority(get_common_item_group(group_list))
                group_list = []                      
    return priority_counter, badge_priority_counter


if __name__ == "__main__":
    time_start = perf_counter_ns()
    file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\puzzle_day_three.txt"
    print(calculate_total_priority(file_loc))
    print(f"Elapsed time : {(perf_counter_ns() - time_start)/1000:.0f} microseconds")

File: Yr2022/src/day_three.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::time::Instant;

use std::collections::HashSet;



fn main() {   
    let start = Instant::now();
    let file_name= "C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\puzzle_day_three.txt";
    let (tot_priority, tot_group_priority) = sum_priorities(file_name);
    //let other_fast = other_faster_way_maybe();
    let duration = start.elapsed();
    println!("Total Priority : {}, Total Group Priority : {}", tot_priority, tot_group_priority);    
    //println!("Other faster way {}", other_fast);
    println!("Time taken to execute: {} microseconds", duration.as_micros());    
}


fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}


fn get_common_item(rucksack: String) -> char {
    let set_one: HashSet<char> = rucksack[0..rucksack.len()/2].chars().collect() ;
    let set_two: HashSet<char> = rucksack[rucksack.len()/2..rucksack.len()].chars().collect() ;    
    let inter: HashSet<char> = set_one.intersection(&set_two).copied().collect();
    let elem =  inter.iter().next().unwrap().clone();
    return elem; 
}

fn get_common_badge(group_vec: &Vec<String>) -> char {
    let set_one: HashSet<char> = group_vec[0].chars().collect();
    let set_two: HashSet<char> = group_vec[1].chars().collect();
    let set_three: HashSet<char> = group_vec[2].chars().collect();
    let first_inter: HashSet<char> = set_one.intersection(&set_two).copied().collect();
    let second_inter: HashSet<char> = first_inter.intersection(&set_three).copied().collect();
    let elem =  second_inter.iter().next().unwrap().clone();
    return elem; 
}

fn get_priority(input_char : u32) -> u32 {
    match input_char {
        97..=122 => return input_char - 96,
        _ => return input_char - 38,
    }
}


fn sum_priorities(my_file_name: &str) -> (u32, u32) {
    let mut total_priority: u32 = 0;
    let mut total_group_priority: u32 = 0;
    let mut group_vec: Vec<String> = Vec::new();
    if let Ok(lines) = read_lines(my_file_name) {
        for line in lines {
            if let Ok(ip) = line {
                let other_line = ip.clone();
                total_priority += get_priority(get_common_item(ip) as u32);
                group_vec.push(other_line);
                if group_vec.len() == 3 {
                    total_group_priority += get_priority(get_common_badge(&group_vec) as u32);
                    group_vec.clear();
                }               
            }
        }
    }
    return (total_priority, total_group_priority)
}


fn other_faster_way_maybe() -> u32 {
    
    let faster_part_one : u32 = include_str!("C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\puzzle_day_three.txt")
        .split("\n")
        .filter(|x| !x.is_empty())
        .map(|rucksack| {
            let set_one: HashSet<char> = rucksack[0..rucksack.len()/2].chars().collect() ;
            let set_two: HashSet<char> = rucksack[rucksack.len()/2..rucksack.len()].chars().collect() ;    
            let inter: HashSet<char> = set_one.intersection(&set_two).copied().collect();
            let elem = inter.iter().next().unwrap().clone() as u32;
            match elem {
                97..=122 => return elem - 96,
                _ => return elem - 38,
            }
        })
        .sum();

    return faster_part_one;
}